frame/name-service/src/lib.rs

```diff
@@ -0,0 +1,487 @@
+		_,
+		Blake2_128Concat,
+		NameHash,
+		Registration<T::AccountId, BalanceOf<T>, T::BlockNumber>,
+	>;
+
+	/// This resolver maps name hashes to an account
+	#[pallet::storage]
+	#[pallet::getter(fn resolve)]
+	pub(super) type Resolvers<T: Config> = StorageMap<_, Blake2_128Concat, NameHash, T::AccountId>;
+
+	// Your Pallet's events.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		// TODO: Potentially make events more lightweight
+		/// A new `Commitment` has taken place.
+		Committed { sender: T::AccountId, who: T::AccountId, hash: CommitmentHash },
+		/// A new `Registration` has taken added.
+		NameRegistered { name_hash: NameHash, owner: T::AccountId, expiry: T::BlockNumber },
+		/// A `Registration` has been transferred to a new owner.
+		Transfer { from: T::AccountId, to: T::AccountId },
+		/// A `Registration` has been renewed.
+		NameRenewed { name_hash: NameHash, expires: T::BlockNumber },
+		/// An address has been set for a name hash to resolve to.
+		AddressSet { name_hash: NameHash, address: T::AccountId },
+		/// An address was deregistered.
+		AddressDeregistered { name_hash: NameHash },
+	}
+
+	// Your Pallet's error messages.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// This commitment hash already exists in storage.
+		AlreadyCommitted,
+		/// It has not passed the minimum waiting period to reveal a commitement.
+		TooEarlyToReveal,
+		/// The commitment cannot yet be removed. Has not expired.
+		CommitmentNotExpired,
+		/// This commitment does not exist.
+		CommitmentNotFound,
+		/// This registration does not exist.
+		RegistrationNotFound,
+		/// The sender is not the registration owner.
+		NotRegistrationOwner,
+		/// This registration has expired.
+		RegistrationExpired,
+		/// This registration has not yet expired.
+		RegistrationNotExpired,
+	}
+
+	// Your Pallet's callable functions.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		// TODO: Should we allow registration on behalf of?
+		#[pallet::weight(0)]
+		pub fn commit(
+			origin: OriginFor<T>,
+			who: T::AccountId,
+			commitment_hash: CommitmentHash,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			ensure!(!Commitments::<T>::contains_key(commitment_hash), Error::<T>::AlreadyCommitted);
+
+			let block_number = frame_system::Pallet::<T>::block_number();
+			let deposit = T::CommitmentDeposit::get();
+
+			T::Currency::reserve(&sender, deposit)?;
+
+			let commitment = Commitment { who: who.clone(), when: block_number, deposit };
+
+			Commitments::<T>::insert(commitment_hash, commitment);
+			Self::deposit_event(Event::<T>::Committed { sender, who, hash: commitment_hash });
+
+			Ok(())
+		}
+
+		#[pallet::weight(0)]
+		pub fn reveal(
+			origin: OriginFor<T>,
+			name: Vec<u8>,
+			secret: u64,
+			periods: u32,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+			let commitment_hash = sp_io::hashing::blake2_256(&(&name, secret).encode());
+
```

Comment:
Instead of using just a plain blake2 we may need to already consider sub-domains, otherwise someone registers `something.alice.dot` to troll `alice.dot`.  
See ENS recursive sub-domain hashing: https://docs.ens.domains/contract-api-reference/name-processing#algorithm

Reply:
this is handled by the front-end, not the pallet. Show me the code in the contract where they handle this.

Reply:
They have this [setSubnodeOwner](https://github.com/ensdomains/ens/blob/master/contracts/ENSRegistry.sol#L74) function which checks ownership of the parent node, explained here: https://docs.ens.domains/contract-api-reference/ens#set-subdomain-owner.  
I think this is preventing someone else than *Alice* to buy `sub.alice.dot`.

Reply:
We follow the same hashing pattern for subnodes.

If I'm following your concern correctly, only the owner of `alice.dot` would be able to register `something.alice.dot`. If an attempt is made to register `something.alice.dot` directly it would fail at multiple points - parent does not exist, and not the current owner of parent.